File: server/src/training/datasetPaths.ts
import { createHash } from 'node:crypto';
import { existsSync } from 'node:fs';
import { mkdir, readFile, writeFile } from 'node:fs/promises';
import { dirname, resolve } from 'node:path';
import {
  resolveBioDatasetRoot,
  resolveGoldDatasetRoot,
  resolveRepositoryRoot,
  resolveStyleGoldOutputPath,
} from '../runtime/artifactPaths.js';

export interface DatasetMirrorManifestInput {
  source: string;
  mirror: string;
  datasetVersion: string;
  createdAt?: string;
}

export interface ResolvedDatasetPath {
  path: string;
  source: 'engine-v2' | 'legacy';
  warning: string | null;
}

function readOverride(name: string): string | null {
  const value = process.env[name]?.trim();
  return value ? value : null;
}

function warnLegacyDatasetFallback(message: string): void {
  console.warn(`[dataset-paths] ${message}`);
}

function resolveFirstExisting(
  engineV2Path: string,
  legacyPath: string,
  description: string,
): ResolvedDatasetPath {
  if (existsSync(engineV2Path)) {
    return {
      path: engineV2Path,
      source: 'engine-v2',
      warning: null,
    };
  }

  if (existsSync(legacyPath)) {
    const warning = `${description} is falling back to legacy path: ${legacyPath}`;
    warnLegacyDatasetFallback(warning);
    return {
      path: legacyPath,
      source: 'legacy',
      warning,
    };
  }

  return {
    path: engineV2Path,
    source: 'engine-v2',
    warning: null,
  };
}

export function resolveEngineV1DatasetRoot(): string {
  return readOverride('BULKREFERENCES_ENGINE_V1_DATASET_ROOT')
    ?? resolve(resolveRepositoryRoot(), 'datasets', 'engine-v1');
}

export function resolveEngineV2DatasetRoot(): string {
  return readOverride('BULKREFERENCES_ENGINE_V2_DATASET_ROOT')
    ?? resolve(resolveRepositoryRoot(), 'datasets', 'engine-v2');
}

export function resolveEngineV2StyleCoreRoot(): string {
  return resolve(resolveEngineV2DatasetRoot(), 'gold', 'style-core');
}

export function resolveEngineV2StyleCoreCuratedRoot(): string {
  return resolve(resolveEngineV2StyleCoreRoot(), 'curated');
}

export function resolveEngineV2StyleCoreQuarantineRoot(): string {
  return resolve(resolveEngineV2StyleCoreRoot(), 'quarantine');
}

export function resolveEngineV2StyleCoreSourcePacksRoot(): string {
  return resolve(resolveEngineV2StyleCoreRoot(), 'source-packs');
}

export function resolveEngineV2StyleCoreExportsRoot(): string {
  return resolve(resolveEngineV2StyleCoreRoot(), 'exports');
}

export function resolveEngineV2StyleGoldExportPath(): string {
  return readOverride('BULKREFERENCES_ENGINE_V2_STYLE_GOLD_EXPORT_PATH')
    ?? resolve(resolveEngineV2StyleCoreExportsRoot(), 'style_gold.jsonl');
}

export function resolveEngineV2CitationBioRoot(): string {
  return readOverride('BULKREFERENCES_ENGINE_V2_CITATION_BIO_ROOT')
    ?? resolve(resolveEngineV2DatasetRoot(), 'gold', 'citation-bio');
}

export function resolveEngineV2AuthorityRoot(): string {
  return readOverride('BULKREFERENCES_ENGINE_V2_AUTHORITY_ROOT')
    ?? resolve(resolveEngineV2DatasetRoot(), 'gold', 'authority');
}

export function resolveStyleGoldSourcePath(): ResolvedDatasetPath {
  return resolveFirstExisting(
    resolveEngineV2StyleGoldExportPath(),
    resolveStyleGoldOutputPath(),
    'style gold export',
  );
}

export function resolveGoldDatasetSourceRoot(): ResolvedDatasetPath {
  return resolveFirstExisting(
    resolveEngineV2StyleCoreCuratedRoot(),
    resolveGoldDatasetRoot(),
    'style-core curated dataset root',
  );
}

export function resolveCitationBioSourceRoot(): ResolvedDatasetPath {
  return resolveFirstExisting(
    resolveEngineV2CitationBioRoot(),
    resolveBioDatasetRoot(),
    'citation BIO dataset root',
  );
}

export async function writeDatasetMirrorManifest(
  input: DatasetMirrorManifestInput,
): Promise<string> {
  const sourceContent = await readFile(input.source);
  const manifest = {
    source: input.source,
    mirror: input.mirror,
    dataset_version: input.datasetVersion,
    created_at: input.createdAt ?? new Date().toISOString(),
    sha256: createHash('sha256').update(sourceContent).digest('hex'),
  };
  const manifestPath = `${input.mirror}.manifest.json`;
  await mkdir(dirname(manifestPath), { recursive: true });
  await writeFile(manifestPath, `${JSON.stringify(manifest, null, 2)}\n`, 'utf8');
  return manifestPath;
}

File: ml-service/app/dataset_paths.py
from __future__ import annotations

import os
import warnings
from pathlib import Path


ML_SERVICE_ROOT = Path(__file__).resolve().parents[1]
REPOSITORY_ROOT = ML_SERVICE_ROOT.parent


def _read_override(name: str) -> Path | None:
    value = os.environ.get(name, "").strip()
    return Path(value).expanduser().resolve() if value else None


def _warn_legacy_fallback(description: str, path: Path) -> None:
    warnings.warn(
        f"[dataset-paths] {description} is falling back to legacy path: {path}",
        RuntimeWarning,
        stacklevel=2,
    )


def engine_v2_style_gold_export_path() -> Path:
    return _read_override("BULKREFERENCES_ENGINE_V2_STYLE_GOLD_EXPORT_PATH") or (
        REPOSITORY_ROOT
        / "datasets"
        / "engine-v2"
        / "gold"
        / "style-core"
        / "exports"
        / "style_gold.jsonl"
    )


def legacy_style_gold_export_path() -> Path:
    return _read_override("BULKREFERENCES_STYLE_GOLD_OUTPUT_PATH") or (
        ML_SERVICE_ROOT / "training" / "style_gold.jsonl"
    )


def engine_v2_citation_bio_root() -> Path:
    return _read_override("BULKREFERENCES_ENGINE_V2_CITATION_BIO_ROOT") or (
        REPOSITORY_ROOT / "datasets" / "engine-v2" / "gold" / "citation-bio"
    )


def legacy_citation_bio_root() -> Path:
    return _read_override("BULKREFERENCES_BIO_DATASET_ROOT") or (
        ML_SERVICE_ROOT / "datasets" / "citation-bio"
    )


def resolve_training_jsonl_path(path: str | Path | None) -> Path:
    if path is None:
        v2_path = engine_v2_style_gold_export_path()
        if v2_path.exists():
            return v2_path
        legacy_path = legacy_style_gold_export_path()
        if legacy_path.exists():
            _warn_legacy_fallback("style gold export", legacy_path)
            return legacy_path
        return v2_path

    requested_path = Path(path)
    if requested_path.exists():
        return requested_path

    legacy_path = legacy_style_gold_export_path()
    if requested_path.resolve() == legacy_path.resolve():
        v2_path = engine_v2_style_gold_export_path()
        if v2_path.exists():
            return v2_path

    return requested_path


def resolve_bio_gold_jsonl_path(path: str | Path) -> Path:
    requested_path = Path(path)
    if requested_path.exists():
        return requested_path

    legacy_root = legacy_citation_bio_root().resolve()
    try:
        relative_path = requested_path.resolve().relative_to(legacy_root)
    except ValueError:
        return requested_path

    v2_path = engine_v2_citation_bio_root() / relative_path
    if v2_path.exists():
        return v2_path

    return requested_path
